src/runner_metrics.py

```diff
@@ -0,0 +1,222 @@
+
+    Args:
+        fs: The shared filesystem for a specific runner.
+
+    Returns:
+        A tuple of files whose size is larger than the limit.
+    """
+    files: list[Path] = [
+        fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(POST_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME),
+    ]
+
+    return tuple(
+        filter(lambda file: file.exists() and file.stat().st_size > FILE_SIZE_BYTES_LIMIT, files)
+    )
+
+
+def _extract_metrics_from_fs(fs: shared_fs.SharedFilesystem) -> Optional[RunnerMetrics]:
+    """Extract metrics from a shared filesystem.
+
+    Args:
+        fs: The shared filesystem for a specific runner.
+
+    Returns:
+        The extracted metrics if at least the pre-job metrics are present.
+
+    Raises:
+        CorruptMetricDataError: Raised if one of the files is not valid or too large.
+    """
+    if too_large_files := _inspect_file_sizes(fs):
+        raise CorruptMetricDataError(
+            f"File size of {too_large_files} is too large. "
+            f"The limit is {FILE_SIZE_BYTES_LIMIT} bytes."
+        )
+
+    try:
+        installed_timestamp = fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME).read_text()
+    except FileNotFoundError:
+        logger.exception("installed_timestamp not found for runner %s", fs.runner_name)
+        return None
+
+    logger.debug("Runner %s installed at %s", fs.runner_name, installed_timestamp)
+    try:
+        pre_job_metrics = json.loads(fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME).read_text())
+        logger.debug("Pre-job metrics for runner %s: %s", fs.runner_name, pre_job_metrics)
+    except FileNotFoundError:
+        logger.warning("%s not found for runner %s.", PRE_JOB_METRICS_FILE_NAME, fs.runner_name)
+        return None
+    except JSONDecodeError as exc:
+        raise CorruptMetricDataError(str(exc)) from exc
+
+    try:
+        post_job_metrics = json.loads(fs.path.joinpath(POST_JOB_METRICS_FILE_NAME).read_text())
+        logger.debug("Post-job metrics for runner %s: %s", fs.runner_name, post_job_metrics)
+    except FileNotFoundError:
+        logger.warning("%s not found for runner %s", POST_JOB_METRICS_FILE_NAME, fs.runner_name)
+        post_job_metrics = None
+    except JSONDecodeError as exc:
+        raise CorruptMetricDataError(str(exc)) from exc
+
+    if not isinstance(pre_job_metrics, dict):
+        raise CorruptMetricDataError(
+            f"Pre-job metrics for runner {fs.runner_name} is not a JSON object."
+        )
+
+    if not isinstance(post_job_metrics, dict) and post_job_metrics is not None:
+        raise CorruptMetricDataError(
+            f"Post-job metrics for runner {fs.runner_name} is not a JSON object."
+        )
+
+    try:
+        return RunnerMetrics(
+            installed_timestamp=installed_timestamp,
+            pre_job=PreJobMetrics(**pre_job_metrics),
+            post_job=PostJobMetrics(**post_job_metrics) if post_job_metrics else None,
```

Comment:
When are the PostJobMetrics used ? I can't find any reference to it

Reply:
Not used yet. This will be used in a follow-up PR where all metrics (including `RunnerStop`) will be logged.